rules/TOOL_trace_nsiqcppstyle_callbacks.py

```diff
@@ -0,0 +1,80 @@
+from nsiqcppstyle_rulemanager import *
+
+def PrintContextStack(contextStack):
+    if (contextStack != None) and (contextStack.contextstack != None) and (len(contextStack.contextstack) > 0):
+        indent_string = (20 * ' ')
+        print("%sContext stack (%d):" % (indent_string, len(contextStack.contextstack)))
+        for t in contextStack.contextstack:
+            print('%s    %s' % (indent_string, str(t)))
+
+def FunctionScopeRule(lexer, contextStack):
+    print("FunctionScopeRule (lexer, contextStack)")
+    PrintContextStack(contextStack)
+
+def FunctionNameRule(lexer, fullName, decl, contextStack, context):
+    print("FunctionNameRule  (lexer,")
+    print("                   fullName='%s'," % (fullName))
+    print("                   decl='%s'," % (decl))
+    print("                   contextStack,")
+    print("                   context='%s')" % (str(context)))
+    PrintContextStack(contextStack)
+
+
+def PreprocessRule(lexer, contextStack):
+    print("PreprocessRule    (lexer,")
+    print("                   contextStack,")
+    print("                   token=%s" % (lexer.GetCurToken()))
+    PrintContextStack(contextStack)
+
+def CommentRule(lexer, token):
+    print("CommentRule       (lexer, token=%s)" % (lexer.GetCurToken()))
+
+def LineRule(lexer, line, lineNumber):
+    print("--------------------------------------------------")
+    print("LineRule          (lexer, line='%s', lineNumber=%d)" % (line, lineNumber))
+
+def TokenRule(lexer, contextStack):
+    print("TokenRule         (lexer,")
+    print("                   contextStack,")
+    print("                   token=%s)" % (lexer.GetCurToken()))
+    PrintContextStack(contextStack)
+
+def FileStartRule(lexer, filename, dirname):
+    print("FileStartRule     (lexer, filename='%s', dirname='%s')" % (filename, dirname))
+
+def FileEndRule(lexer, filename, dirname):
+    print("FileEndRule       (lexer, filename='%s', dirname='%s')" % (filename, dirname))
+
+def ProjectRule(targetName):
+    print("ProjectRule       (targetName='%s')" % (targetName))
+
+def TypeNameRule(lexer, typeName, typeFullName, decl, contextStack, typeContext):
+    print("TypeNameRule      (lexer, typeName='%s', typeFullName='%s', decl='%s')" % (typeName, typeFullName, decl))
+    PrintContextStack(contextStack)
+
+def TypeScopeRule(lexer, contextStack):
+    print("TypeScopeRule     (lexer, contextStack)")
+    PrintContextStack(contextStack)
+
+def SessionStartRule():
+    # Print instructions, legend, etc. to help the developer
+    print("Legend: LexToken contain six fields: (type, value, lineno, column, lexpos, inactive, pp)")
+    print("")
+    print("SessionStartRule  ()")
+
+def SessionEndRule():
+    print("SessionEndRule    ()")
```

Comment:
Could we use `inspect` instead of manually specifying this?

```py
from inspect import stack

def make_verbose(args, alignment: Optional[int] = None):
  caller_name = stack[1].function
  alignment = len(caller_name) if alignment is None else alignment
  print(f"{caller_name:<{alignment}} {args}')
```

The max alignment can be either hardcoded (with appropriate comment) or found dynamically through
```py
  module = inspect.getmodule(make_verbose)
  max_alignment = max(len(x.__name__) for x in module.__dict__.values() if inspect.isfunction(x))
```

The main aim is to reduce content in the function and potentially, move towards a decorator. However, please don't implement a decorator just yet since that would get complicated

Reply:
I originally investigated using dynamic inspection (and computing the maximum alignment factor) but decided not to because:

- The resulting code was more complex and less readable
- The TOOL is not a part of the NsiqCppStyle engine, but rather a very simple, externally-written "rule"-like script.  It does not need to conform to the latest Python "fad", however tempting it may be to do so.

I am aware that the [Zen of Python](https://peps.python.org/pep-0020/) can be interpreted either way, but I feel the conciseness of code does not justify the reduced code readability and increased maintenance complexity.

Having said that, I will investigate using `stack.inspect` as I would like to "wet my feet" with it.

Reply:
I've played with your sample code above.  I must say it is most ingenious!  I modified the code slightly from:

```
max_alignment = max(len(x.__name__) for x in module.__dict__.values() if inspect.isfunction(x))
```

to:

```
max_alignment = max(len(x.__name__) for x in module.__dict__.values() if inspect.isfunction(x)) and x.__name__.endswith("Rule")
```
This was to only use the callback function names in the TOOL.  So getting the maximum callback function name length, then subtracting len("Rule") gives the alignment for any callback function trace.

Having said that, the tool has different types of output lines, each needing alignment (plus/minus 1).  Again, IMHO it is overkill to use anything other than good ol' spaces to align the outputs.  This is a tool/rule, not the NsiqCppStyle engine.  The former does not need to be at the cutting edge of Pythonic coding, mainly keeping it simple and readable.  Had it been the latter, where generic programming is important, I could see using uber-clever coding.

So you need to make the call.  If you still feel it requires state-of-the-art code, then I request that you make the changes after it's been merged in.